tests/subsys/emds/emds_api/src/main.c

```diff
@@ -0,0 +1,302 @@
+ */
+
+#include <stdint.h>
+#include <ztest.h>
+#include <emds/emds.h>
+#if CONFIG_SETTINGS
+#include <settings/settings.h>
+#include <sys/reboot.h>
+#endif
+
+enum test_states {
+	EMDS_TS_EMPTY_FLASH,
+	EMDS_TS_STORE_DATA,
+	EMDS_TS_CLEAR_FLASH,
+	EMDS_TS_NO_STORE
+};
+
+static int iteration;
+
+static enum test_states state[] = {
+	EMDS_TS_EMPTY_FLASH,
+	EMDS_TS_STORE_DATA,
+	EMDS_TS_CLEAR_FLASH,
+	EMDS_TS_EMPTY_FLASH,
+	EMDS_TS_NO_STORE,
+	EMDS_TS_EMPTY_FLASH,
+	EMDS_TS_CLEAR_FLASH,
+};
+
+uint8_t expect_d_data[3][10] = {
+	{ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10},
+	{11, 12, 13, 14, 15, 16, 17, 18, 19, 20},
+	{21, 22, 23, 24, 25, 26, 27, 28, 29, 30},
+};
+
+uint8_t d_data[3][10];
+struct emds_entry d_entries[3] = {
+	{0x1001, &d_data[0][0], 10},
+	{0x1002, &d_data[1][0], 10},
+	{0x1003, &d_data[2][0], 10},
+};
+
+uint8_t expect_s_data[1024] = { 0xCC };
+uint8_t s_data[1024];
+
+EMDS_STATIC_ENTRY_DEFINE(s_entry, 0x100, s_data, sizeof(s_data));
+
+char *print_state(enum test_states s)
+{
+	switch (s) {
+	case EMDS_TS_EMPTY_FLASH:
+		return "EMPTY_FLASH";
+	case EMDS_TS_STORE_DATA:
+		return "STORE_DATA";
+	case EMDS_TS_CLEAR_FLASH:
+		return "CLEAR_FLASH";
+	case EMDS_TS_NO_STORE:
+	default:
+		return "NO_STORE";
+	}
+}
+
+/** Mocks ******************************************/
+
+
+/** End Mocks **************************************/
+
+static void test_init(void)
+{
+	int err;
+
+	err = emds_entry_add(&d_entries[0]);
+	zassert_equal(err, -ECANCELED, "List not initialized");
+
+	err = emds_prepare();
+	zassert_equal(err, -ECANCELED, "Flash not initialized");
+
+	err = emds_clear();
+	zassert_equal(err, -ECANCELED, "Flash not initialized");
+
+	zassert_false(emds_init(), "Initializing failed");
+	zassert_equal(emds_init(), -EALREADY, "Initialization did not fail");
+}
+
+static void test_add_d_entries(void)
+{
+	int err;
+	uint32_t store_expected = NRFX_CEIL_DIV(sizeof(s_data), 4) * 4 + 8;
+
+	for (int i = 0; i < ARRAY_SIZE(d_entries); i++) {
+		err = emds_entry_add(&d_entries[i]);
+		store_expected += NRFX_CEIL_DIV(d_entries[i].len, 4) * 4;
+		store_expected += 8;
+		zassert_equal(err, 0, "Add entry failed");
+
+		err = emds_entry_add(&d_entries[i]);
+		zassert_equal(err, -EINVAL, "Entry duplicated");
+	}
+
+	uint32_t store_used = emds_store_size_get();
+
+	zassert_equal(store_used, store_expected, "Wrong storage size");
+}
+
+static void test_load_empty_flash(void)
+{
+	int err;
+	uint8_t empty_d_data[sizeof(d_data)];
+	uint8_t empty_s_data[sizeof(s_data)];
+
+	memset(empty_d_data, 0, sizeof(d_data));
+	memset(empty_s_data, 0, sizeof(s_data));
+
+	err = emds_load();
+	zassert_equal(err, 0, "Load failed");
+
+	zassert_mem_equal(d_data, empty_d_data, sizeof(d_data),
+			  "Data has changed");
+	zassert_mem_equal(s_data, empty_s_data, sizeof(s_data),
+			  "Data has changed");
+}
+
+static void test_load_flash(void)
+{
+	int err;
+
+	err = emds_load();
+	zassert_equal(err, 0, "Load failed");
```

Comment:
```suggestion
	zassert_equal(emds_load(), 0, "Load failed");
```
Nitpick/matter of opinion: Is it necessary to create a err variable? Applies for several places.